src/javascript/app_2/Stores/base_store.js

```diff
@@ -0,0 +1,139 @@
+     * @param {String[]} session_storage_properties - A list of properties' names that should be kept in sessionStorage.
+     */
+    constructor(
+        root_store = null,
+        local_storage_properties = null, 
+        session_storage_properties = null
+    ) {
+
+        Object.defineProperty(this, 'root_store', {
+            enumerable: false,
+            writable  : true,
+        });
+        Object.defineProperty(this, 'local_storage_properties', {
+            enumerable: false,
+            writable  : true,
+        });
+        Object.defineProperty(this, 'session_storage_properties', {
+            enumerable: false,
+            writable  : true,
+        });
+
+        this.root_store = root_store;
+        this.local_storage_properties   = local_storage_properties || [];
+        this.session_storage_properties = session_storage_properties || [];
+        
+        this.setupReactionForLocalStorage();
+        this.setupReactionForSessionStorage();
+        this.retrieveFromStorage();
+    }
+
+    /**
+     * Returns an snapshot to the current store
+     *
+     * @param {String[]} properties - A list of properties' names that should be in the snapshot.
+     *
+     * @return {Object} Returns a clone object for the store.
+     */
+    getSnapshot(properties) {
+        let snapshot = toJS(this);
+
+        if (!isEmptyObject(this.root_store)) {
+            snapshot.root_store = this.root_store;
+        }
+
+        if (properties && properties.length) {
+            snapshot = properties.reduce(
+                (result, p) => Object.assign(result, { [p]: snapshot[p]}),
+                {}
+            );
+        }
+
+        return snapshot;
+    }
+
+    /**
+     * Setups a reaction on properties which are mentioned in `local_storage_properties`
+     *  and invoke `saveToStorage` when there are any changes on them.
+     *
+     */
+    setupReactionForLocalStorage() {
+        if (this.local_storage_properties.length) {
+            reaction(
+                () => this.local_storage_properties.map(i => this[i]),
+                () => this.saveToStorage(this.local_storage_properties, BaseStore.STORAGES.LOCAL_STORAGE)
+            );
+        }
+    }
+
+    /**
+     * Setups a reaction on properties which are mentioned in `session_storage_properties`
+     *  and invoke `saveToStorage` when there are any changes on them.
+     *
+     */
+    setupReactionForSessionStorage() {
+        if (this.session_storage_properties.length) {
+            reaction(
+                () => this.local_storage_properties.map(i => this[i]),
+                () => this.saveToStorage(this.session_storage_properties, BaseStore.STORAGES.SESSION_STORAGE)
+            );
+        }
+    }
+
+    /**
+     * Saves the snapshot of the store includes properties that are mentioned in
+     *  first parameter in the storage that is mentioned in second parameters.
```

Comment:
this whole sentence is confusing 😅 maybe:
`Store the passed properties in the particular storage and save a snapshot of it`

Reply:
It doesn't store the properties in the particular storage. Actually, it gets a snapshot of the store and just removes properties that aren't mentioned in `properties` parameters from the snapshot and then save it to the particular storage.

Reply:
I see! maybe summarise that and use it instead? 😝

Reply:
if you think the summarised one conveys the similar meaning, I'll change it.

Reply:
I think the comment right now only mention `includes`, but your explanation is more like it uses those properties to decide which to remove, so maybe something like
`Removes properties that are not passed from the snapshot of the store and saves it to the passed storage`

Reply:
if you can think of a clearer explanation please use that.